fix: bug fix for obs_sort_module.F90.

greent12 opened the pull request:
This is a proposed fix for Issue #11 

A conditional statement was added on line 3158 to check wether the pointer 'next' is associated or not before going into the following block of code. This bug fix was stated by the user "Alpha_su" in the obsgrid forum: https://forum.mmm.ucar.edu/threads/obsgrid-exe-fails-after-a-few-seconds.13474/

greent12 committed:
fix: bug fix for obs_sort_module.F90. Pointer 'next' on line 3160 (formerly 3159) was causing obsgrid to segfault situationally.
A conditional statement was added on line 3158 to check wether the pointer 'next' is associated or not before going into the following block of code. This seems to have remedied the problem for now.
This bug fix was stated by the user "Alpha_su" in the obsgrid forum: https://forum.mmm.ucar.edu/threads/obsgrid-exe-fails-after-a-few-seconds.13474/

brianreen commented:
Thanks for submitting this code to address issue #11.  The original code inadvertently relies on short-circuiting of Boolean expressions, but since the Fortran standard does not require this, the code only works with certain compilers. While the fix you submitted would avoid the issue, it seems to me that it would be more straightforward to not add the `ASSOCIATED` check but rather to break the existing `IF` statement into two `IF` statements so that `next` is evaluated only when `is_sounding` is `FALSE.`  Specifically, I would suggest:
```
        IF(.NOT. is_sounding) THEN
          IF(.NOT. eps_equal(next%meas%height%data, obs(i)%info%elevation, 1.)) THEN
            obs(i)%info%discard = .TRUE.
            PRINT '(A,A,A,A,F12.2,A,F12.2,A,A,A,A)','WARNING: Ob indicates is_sounding=.FALSE. ',&
             'but the ob height does ',&
             'not match the station elevation.  This ob will be omitted since Obsgrid assumes ',&
             'is_sounding=.FALSE. indicates a surface ob. Height = ', next%meas%height%data, &
             ', Elevation = ', obs(i)%info%elevation,', Ob ID = ',TRIM(ADJUSTL(obs(i)%location%id)),&
             ', Ob Time = ', obs(i)%valid_time%date_char
          ENDIF
        ENDIF
```

greent12 committed:
Redo of bug fix for ops_sort_module.F90

Changes made to obs_sort_module.F90 based on suggestion to restructure conditional statements.

greent12 commented:
Thanks for the suggestion @brianreen. I made that change and you'll see it as another commit for this pull request. 
I think it's still important to maintain OBSGRID, as it's still a frequently used tool in the WRF community (see the recent comments in https://forum.mmm.ucar.edu/threads/obsgrid-exe-fails-after-a-few-seconds.13474/). Although it's a small issue to fix, I think making the source code stable for any compiler is still important. Let me know if I need to do anything else to try to get this pull request done.
